src/Renci.SshNet/PrivateKeyFile.PuTTY.cs

```diff
@@ -0,0 +1,273 @@
+                                throw new SshException("PuTTY key file version " + _version + " is not supported");
+                        }
+
+                        using (var cipher = new AesCipher(cipherKey, cipherIV, AesCipherMode.CBC, pkcs7Padding: false))
+                        {
+                            privateKey = cipher.Decrypt(_data);
+                        }
+
+                        break;
+                    case "none":
+                        switch (_version)
+                        {
+                            case "3":
+                                hmac = new HMACSHA256(Array.Empty<byte>());
+                                break;
+                            case "2":
+                                var macKey = CryptoAbstraction.HashSHA1(Encoding.UTF8.GetBytes("putty-private-key-file-mac-key"));
+                                hmac = new HMACSHA1(macKey);
+                                break;
+                            default:
+                                throw new SshException("PuTTY key file version " + _version + " is not supported");
+                        }
+
+                        privateKey = _data;
+                        break;
+                    default:
+                        throw new SshException("Encryption " + _encryptionType + " is not supported for PuTTY key file");
+                }
+
+                byte[] macData;
+                using (var macStream = new SshDataStream(256))
+                {
+                    macStream.Write(_algorithmName, Encoding.UTF8);
+                    macStream.Write(_encryptionType, Encoding.UTF8);
+                    macStream.Write(_comment, Encoding.UTF8);
+                    macStream.WriteBinary(_publicKey);
+                    macStream.WriteBinary(privateKey);
+                    macData = macStream.ToArray();
+                }
+
+                byte[] macValue;
+                using (hmac)
+                {
+                    macValue = hmac.ComputeHash(macData);
+                }
+#if NET
+                var reference = Convert.FromHexString(_mac);
+#else
+                var reference = Org.BouncyCastle.Utilities.Encoders.Hex.Decode(_mac);
+#endif
+                if (!macValue.SequenceEqual(reference))
+                {
+                    throw new SshException("MAC verification failed for PuTTY key file");
+                }
+
+                var publicKeyReader = new SshDataReader(_publicKey);
+                var keyType = publicKeyReader.ReadString(Encoding.UTF8);
+                Debug.Assert(keyType == _algorithmName, $"{nameof(keyType)} is not the same with {nameof(_algorithmName)}");
+
+                var privateKeyReader = new SshDataReader(privateKey);
+
+                Key parsedKey;
+
+                switch (keyType)
+                {
+                    case "ssh-ed25519":
+                        parsedKey = new ED25519Key(privateKeyReader.ReadBignum2());
```

Comment:
`ED25519Key` class does not have a constructor with **both** `publicKeyData` and `privateKeyData`. The current implementation is to generate public key from private key. Should we add a new constructor?
https://github.com/sshnet/SSH.NET/blob/3ec45e1af31c84a184c518741d0542321d6deedf/src/Renci.SshNet/Security/Cryptography/ED25519Key.cs#L98-L104

Reply:
it seems ok as it is, but I don't mind